README.md
# PyTorch-SM3
[[source](https://arxiv.org/abs/1901.11150)]
[[TensorFlow](https://github.com/google-research/google-research/tree/master/sm3)]
[[notebook](./SM3_example.ipynb)]

Implements the SM3-II adaptive optimization algorithm for PyTorch.
This algorithm was designed by Rohan Anil, Vineet Gupta, Tomer Koren, and
Yoram Singer and implemented in TensorFlow.

The 'Square-root of Minima of Sums of Maxima of Squared-gradients Method'
(SM3) algorithm is a memory-efficient adaptive optimization algorithm similar
to Adam and Adagrad with greatly reduced memory usage for history tensors.
For an `n x m` matrix, Adam and Adagrad use `O(nm)` memory for history
tensors, while SM3 uses `O(n+m)` due to the chosen cover. In general, a tensor
of shape `(n_1, n_2, ..., n_k)` optimized using Adam will use `O(prod n_i)`
memory for storage tensors, while the optimization using SM3 will use
`O(sum n_i)` memory. Despite storing fewer parameters, this optimization
algorithm manages to be comparably effective.

This advantage drastically shrinks when `momentum > 0`. The momentum is
tracked using a tensor of the same shape as the tensor being optimized. With
momentum, SM3 will use just over half as much memory as Adam, and a bit more
than Adagrad.

If the gradient is sparse, then the optimization algorithm will use `O(n_1)`
memory as there is only a row cover. The value of `momentum` is ignored in
this case.

## Installing
To install with `pip`, you can use `pip install torch-SM3`. Alternatively,
clone the repository and run `python setup.py sdist` and install using the
generated source package.

## Usage
After installing, import the optimizer using `from SM3 import SM3`. The `SM3`
optimizer that is imported can be used exactly the same way a PyTorch
optimizer. For example, the optimizer can be constructed using
`opt = SM3(model.parameters())` with parameter updates being applied using
`opt.step()`.

## Implementation Differences
The algorithm presented by the original authors mentions that the optimization
algorithm can be modified to use exponential moving averages. I incorporated
this into the optimizer. If `beta = 0`, then the accumulated gradient squares
method (i.e. the default SM3 method) is used. If `beta > 0`, then the updates
use exponential moving averages instead. The authors found that `beta = 0` 
was superior for their experiments in translation and language models.

## Requirements
The requirements given in `requirements.txt` are not the absolute minimum -
the optimizer may function for earlier versions of PyTorch than 1.4. However,
these versions are not tested against. Furthermore, a change in the backend
`C++` signatures means that the current version of this package may not run
against earlier versions of PyTorch.

# Wisdom from authors
Their full advice can be seen in the sources above. Here are two points they
emphasize and how to incorporate them.

## Learning rate warm-up
They prefer using a learning rate that quadratically ramps up to the
full learning rate. This is done in the notebook linked above by using the
`LambdaLR` class. After creating the optimizer, you can use the following:
```
lr_lambda = lambda epoch: min(1., (epoch / warm_up_epochs) ** 2)
scheduler = torch.optim.lr_scheduler.LambdaLR(opt, lr_lambda)
```
The authors advocate for this as they found that the early gradients were
typically very large in magnitude. By using a warm-up, the accumulated
gradients are not dominated by the first few updates. After this warm-up,
they do not adjust the learning rate.

## Learning rate decay
[Polyak averaging](https://www.tensorflow.org/api_docs/python/tf/train/ExponentialMovingAverage)
can be useful for training models as the moving average of the parameters
can produce better results than the parameters themselves. As this can be
costly in memory, an alternative they present is to ramp the learning rate
decay to 0 in the last 10% of steps. This can also be achieved using the
`LambdaLR` class with the following `lambda` function:
```
lr_lambda = lambda epoch: min(1., (total_epochs - epoch) / (0.1 * total_epochs))
```
To incorporate both warm-up and decay, we can combine the two functions:
```
lr_lambda = lambda epoch: min(1., (epoch / (warm_up_epochs)) ** 2, (epochs - epoch) / (0.1 * epochs))
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Types of changes
- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.

## Unreleased
### Fixed
- Parameters without a gradient no longer cause an AttributeError exception. This occurs when a model has a parameter with the `requires_grad` attribute set to `False` (e.g. the model has a frozen layer). Previously, the exception would be raised when the SM3 class method `step` is called.

## [0.1.0] - 2020-07-11
### Changed
- Updated optimizer to use new PyTorch function signatures as the previous
signatures were being deprecated. This means prior releases of PyTorch may not
work with this optimizer.
- Updated repository structure for packaging. The code for the optimizer has
been moved to `src/`. The code for testing the optimizer against NumPy has been
moved to `tests/`.
- Updated GitHub workflow to create `pip` packaging and test

## [0.0.0] - 2020-02-04
### Added
- Created `PyTorch-SM3` repository and finalized initial code.


test_SM3.py
import torch
import numpy
import pytest
from SM3 import SM3


def _test_dense_0d_updates(lr, momentum, beta):
    var = numpy.array(0.5)
    grad = numpy.array(0.1)

    x = _create_tensor(var, grad)

    opt = SM3([x], lr=lr, momentum=momentum, beta=beta)
    gbar = numpy.zeros_like(grad)
    accumulator = numpy.zeros_like(grad)
    for _ in range(5):
        # Run a step of training.
        opt.step()

        if beta > 0.:
            accumulator = beta * accumulator + (1. - beta) * numpy.square(grad)
        else:
            accumulator = accumulator + numpy.square(grad)

        exp_p_grad = grad / numpy.sqrt(accumulator)
        gbar = momentum * gbar + (1. - momentum) * exp_p_grad
        var = var - lr * gbar
        # Check that variable and momentum are as expected after one step of
        # training.
        _check_values(x.numpy(), var)
        if momentum > 0.:
            _check_values(opt.state[x]['momentum_buffer'].numpy(), gbar)


def _test_dense_1d_updates(lr, momentum, beta):
    var = numpy.array([0.5, 0.3])
    grad = numpy.array([0.1, 0.1])

    x = _create_tensor(var, grad)

    opt = SM3([x], lr=lr, momentum=momentum, beta=beta)

    gbar = numpy.zeros_like(grad)
    accumulator = numpy.zeros_like(grad)
    for _ in range(5):
        # Run a step of training.
        opt.step()

        if beta > 0.:
            accumulator = beta * accumulator + (1. - beta) * numpy.square(grad)
        else:
            accumulator = accumulator + numpy.square(grad)

        exp_p_grad = grad / numpy.sqrt(accumulator)
        gbar = momentum * gbar + (1. - momentum) * exp_p_grad
        var = var - lr * gbar
        # Check that variable and momentum are as expected after one step of
        # training.
        _check_values(x.numpy(), var)
        if momentum > 0.:
            _check_values(opt.state[x]['momentum_buffer'].numpy(), gbar)

def _test_dense_2d_updates(lr, momentum, beta):
    var = numpy.array([[.5, .5], [.5, .5]])
    grad = numpy.array([[.1, .05], [.02, .03]])

    x = _create_tensor(var, grad)

    opt = SM3([x], lr=lr, momentum=momentum, beta=beta)

    row_accumulator = numpy.zeros([2, 1])
    col_accumulator = numpy.zeros([1, 2])
    gbar = numpy.zeros_like(grad)
    accumulator = numpy.zeros_like(grad)
    for _ in range(5):
        # Run a step of training.
        opt.step()

        accumulator = numpy.minimum(row_accumulator, col_accumulator)
        # Expected preconditioned gradient, momentum, and parameter.
        if beta > 0.:
            accumulator = beta * accumulator + (1. - beta) * numpy.square(grad)
        else:
            accumulator = accumulator + numpy.square(grad)
        # Update SM3 accumulators.
        row_accumulator = numpy.amax(accumulator, axis=1, keepdims=True)
        col_accumulator = numpy.amax(accumulator, axis=0, keepdims=True)
        exp_p_grad = grad / numpy.sqrt(accumulator)
        gbar = momentum * gbar + (1. - momentum) * exp_p_grad
        var = var - lr * gbar
        # Check that variable and momentum are as expected after one step of
        # training.
        _check_values(x.numpy(), var)
        if momentum > 0.:
            _check_values(opt.state[x]['momentum_buffer'].numpy(), gbar)

@pytest.mark.parametrize(
    'lr, beta',
    [
        [0.1, 0.0],
        [0.1, 0.1],
        [0.2, 0.3],
        [0.3, 0.1],
        [0.1, 0.9]
    ]
)
def test_sparse_updates(lr, beta):
    var = numpy.array([[0.5, 0.05], [0.05, 1.0], [0.15, 3.0], [0.35, 2.0]])
    # A sparse gradient that updates index 1, and 3.
    indices = [1, 3]
    grad_values = numpy.array([0.1, 0.05, 0.01, 1.5])
    grad_indices = torch.LongTensor([[1, 1, 3, 3], [0, 1, 0, 1]])
    shape = torch.Size(var.shape)
    sparse_grad = torch.sparse.FloatTensor(
        grad_indices,
        torch.from_numpy(grad_values),
        shape
    )

    x = _create_tensor(var, sparse_grad)
    opt = SM3([x], lr=lr, beta=beta)
    row_accumulator = numpy.zeros([4, 1])

    for _ in range(5):
        opt.step()

        accumulator = numpy.repeat(row_accumulator, 2, 1)
        if beta > 0.:
            accumulator[indices, :] = beta * accumulator[indices, :] \
                + (1. - beta) * numpy.square(grad_values.reshape(2, 2))
        else:
            accumulator[indices, :] += numpy.square(grad_values.reshape(2, 2))

        row_accumulator = numpy.amax(accumulator, axis=1, keepdims=True)
        # Update SM3 accumulators.
        exp_p_grad_values = grad_values.reshape(2, 2) / numpy.sqrt(accumulator[indices, :])
        var[indices, :] = var[indices, :] - lr * exp_p_grad_values

        _check_values(x.numpy()[indices, :], var[indices, :])
        _check_values(opt.state[x]['accumulator_0'].numpy().flatten(), row_accumulator.flatten())

@pytest.mark.parametrize(
    'lr, momentum, beta',
    [
        [0.1, 0.0, 0.0],
        [0.1, 0.1, 0.0],
        [0.1, 0.0, 0.1],
        [0.2, 0.3, 0.1],
        [0.3, 0.1, 0.2],
        [0.1, 0.2, 0.3]
    ]
)
def test_dense_updates(lr, momentum, beta):
    _test_dense_0d_updates(lr, momentum, beta)
    _test_dense_1d_updates(lr, momentum, beta)
    _test_dense_2d_updates(lr, momentum, beta)


def _check_values(var, exp_var):
    assert numpy.isclose(var, exp_var).all()

def _create_tensor(var, grad):
    # var may change independently of x, so clone is necessary.
    x = torch.from_numpy(var).clone()
    if isinstance(grad, torch.Tensor):
        x.grad = grad
    else:
        x.grad = torch.from_numpy(grad)
    return x


tests/test_SM3.py
import torch
import numpy
import pytest
from SM3 import SM3


def test_with_frozen():
    # Reported issue: https://github.com/Enealor/PyTorch-SM3/issues/19
    # This test is designed to fail if an exception is raised when a layer is frozen.

    # Create a small model. The first layer will be frozen.
    test_model = torch.nn.Sequential(
        torch.nn.Linear(10, 5).requires_grad_(False),  # frozen
        torch.nn.Linear(5, 1),  # unfrozen
    )

    # Create optimizer, and input. The LR, and input value can be anything.
    test_optim = SM3(test_model.parameters(), lr=0.1)
    test_input = torch.ones(1, 10)

    # Run an update stage with a loss function. The loss function used can be anything.
    test_optim.zero_grad()
    test_loss = torch.nn.functional.mse_loss(test_model(test_input), torch.zeros(1, 1))
    test_loss.backward()
    # This test fails if an exception is raised. Everything else is a pass.
    try:
        test_optim.step()
    except AttributeError:
        assert False


@pytest.mark.parametrize(
    'lr, beta',
    [
        [0.1, 0.0],
        [0.1, 0.1],
        [0.2, 0.3],
        [0.3, 0.1],
        [0.1, 0.9]
    ]
)
def test_sparse_updates(lr, beta):
    var = numpy.array([[0.5, 0.05], [0.05, 1.0], [0.15, 3.0], [0.35, 2.0]])
    # A sparse gradient that updates index 1, and 3.
    indices = [1, 3]
    grad_values = numpy.array([0.1, 0.05, 0.01, 1.5])
    grad_indices = torch.LongTensor([[1, 1, 3, 3], [0, 1, 0, 1]])
    shape = torch.Size(var.shape)
    sparse_grad = torch.sparse.FloatTensor(
        grad_indices,
        torch.from_numpy(grad_values),
        shape
    )

    x = _create_tensor(var, sparse_grad)
    opt = SM3([x], lr=lr, beta=beta)
    row_accumulator = numpy.zeros([4, 1])

    for _ in range(5):
        opt.step()

        accumulator = numpy.repeat(row_accumulator, 2, 1)
        if beta > 0.:
            accumulator[indices, :] = beta * accumulator[indices, :] \
                + (1. - beta) * numpy.square(grad_values.reshape(2, 2))
        else:
            accumulator[indices, :] += numpy.square(grad_values.reshape(2, 2))

        row_accumulator = numpy.amax(accumulator, axis=1, keepdims=True)
        # Update SM3 accumulators.
        exp_p_grad_values = grad_values.reshape(2, 2) / numpy.sqrt(accumulator[indices, :])
        var[indices, :] = var[indices, :] - lr * exp_p_grad_values

        _check_values(x.numpy()[indices, :], var[indices, :])
        _check_values(opt.state[x]['accumulator_0'].numpy().flatten(), row_accumulator.flatten())


@pytest.mark.parametrize(
    'lr, momentum, beta',
    [
        [0.1, 0.0, 0.0],
        [0.1, 0.1, 0.0],
        [0.1, 0.0, 0.1],
        [0.2, 0.3, 0.1],
        [0.3, 0.1, 0.2],
        [0.1, 0.2, 0.3]
    ]
)
def test_dense_updates(lr, momentum, beta):
    _test_dense_0d_updates(lr, momentum, beta)
    _test_dense_1d_updates(lr, momentum, beta)
    _test_dense_2d_updates(lr, momentum, beta)


def _test_dense_0d_updates(lr, momentum, beta):
    var = numpy.array(0.5)
    grad = numpy.array(0.1)

    x = _create_tensor(var, grad)

    opt = SM3([x], lr=lr, momentum=momentum, beta=beta)
    gbar = numpy.zeros_like(grad)
    accumulator = numpy.zeros_like(grad)
    for _ in range(5):
        # Run a step of training.
        opt.step()

        if beta > 0.:
            accumulator = beta * accumulator + (1. - beta) * numpy.square(grad)
        else:
            accumulator = accumulator + numpy.square(grad)

        exp_p_grad = grad / numpy.sqrt(accumulator)
        gbar = momentum * gbar + (1. - momentum) * exp_p_grad
        var = var - lr * gbar
        # Check that variable and momentum are as expected after one step of
        # training.
        _check_values(x.numpy(), var)
        if momentum > 0.:
            _check_values(opt.state[x]['momentum_buffer'].numpy(), gbar)


def _test_dense_1d_updates(lr, momentum, beta):
    var = numpy.array([0.5, 0.3])
    grad = numpy.array([0.1, 0.1])

    x = _create_tensor(var, grad)

    opt = SM3([x], lr=lr, momentum=momentum, beta=beta)

    gbar = numpy.zeros_like(grad)
    accumulator = numpy.zeros_like(grad)
    for _ in range(5):
        # Run a step of training.
        opt.step()

        if beta > 0.:
            accumulator = beta * accumulator + (1. - beta) * numpy.square(grad)
        else:
            accumulator = accumulator + numpy.square(grad)

        exp_p_grad = grad / numpy.sqrt(accumulator)
        gbar = momentum * gbar + (1. - momentum) * exp_p_grad
        var = var - lr * gbar
        # Check that variable and momentum are as expected after one step of
        # training.
        _check_values(x.numpy(), var)
        if momentum > 0.:
            _check_values(opt.state[x]['momentum_buffer'].numpy(), gbar)


def _test_dense_2d_updates(lr, momentum, beta):
    var = numpy.array([[.5, .5], [.5, .5]])
    grad = numpy.array([[.1, .05], [.02, .03]])

    x = _create_tensor(var, grad)

    opt = SM3([x], lr=lr, momentum=momentum, beta=beta)

    row_accumulator = numpy.zeros([2, 1])
    col_accumulator = numpy.zeros([1, 2])
    gbar = numpy.zeros_like(grad)
    accumulator = numpy.zeros_like(grad)
    for _ in range(5):
        # Run a step of training.
        opt.step()

        accumulator = numpy.minimum(row_accumulator, col_accumulator)
        # Expected preconditioned gradient, momentum, and parameter.
        if beta > 0.:
            accumulator = beta * accumulator + (1. - beta) * numpy.square(grad)
        else:
            accumulator = accumulator + numpy.square(grad)
        # Update SM3 accumulators.
        row_accumulator = numpy.amax(accumulator, axis=1, keepdims=True)
        col_accumulator = numpy.amax(accumulator, axis=0, keepdims=True)
        exp_p_grad = grad / numpy.sqrt(accumulator)
        gbar = momentum * gbar + (1. - momentum) * exp_p_grad
        var = var - lr * gbar
        # Check that variable and momentum are as expected after one step of
        # training.
        _check_values(x.numpy(), var)
        if momentum > 0.:
            _check_values(opt.state[x]['momentum_buffer'].numpy(), gbar)


def _check_values(var, exp_var):
    assert numpy.isclose(var, exp_var).all()


def _create_tensor(var, grad):
    # var may change independently of x, so clone is necessary.
    x = torch.from_numpy(var).clone()
    if isinstance(grad, torch.Tensor):
        x.grad = grad
    else:
        x.grad = torch.from_numpy(grad)
    return x


src/SM3/SM3.py
import torch
from torch.optim import Optimizer

class SM3(Optimizer):
    """Implements SM3 algorithm.

    It has been proposed in `Memory-Efficient Adaptive Optimization`_.

    Arguments:
        params (iterable): iterable of parameters to optimize or dicts defining
            parameter groups
        lr (float, optional): coefficient that scale delta before it is applied
            to the parameters (default: 0.1)
        momentum (float, optional): coefficient used to scale prior updates
            before adding. This drastically increases memory usage if
            `momentum > 0.0`. This is ignored if the parameter's gradient
            is sparse. (default: 0.0)
        beta (float, optional): coefficient used for exponential moving
            averages (default: 0.0)
        eps (float, optional): Term added to square-root in denominator to
            improve numerical stability (default: 1e-30)

    .. _Memory-Efficient Adaptive Optimization:
        https://arxiv.org/abs/1901.11150
    """
    def __init__(self, params, lr=0.1, momentum=0.0, beta=0.0, eps=1e-30):
        if not 0.0 <= lr:
            raise ValueError("Invalid learning rate: {0}".format(lr))
        if not 0.0 <= momentum < 1.0:
            raise ValueError("Invalid momentum: {0}".format(momentum))
        if not 0.0 <= beta < 1.0:
            raise ValueError("Invalid beta: {0}".format(beta))
        if not 0.0 <= eps:
            raise ValueError("Invalid eps: {0}".format(eps))

        defaults = {'lr': lr, 'momentum': momentum, 'beta': beta, 'eps': eps}
        super(SM3, self).__init__(params, defaults)

    @torch.no_grad()
    def step(self, closure=None):
        """Performs a single optimization step.

        Arguments:
            closure (callable, optional): A closure that reevaluates the model
                and returns the loss.
        """
        loss = None
        if closure is not None:
            with torch.enable_grad():
                loss = closure()

        for group in self.param_groups:
            momentum = group['momentum']
            beta = group['beta']
            eps = group['eps']
            for p in group['params']:
                grad = p.grad
                if p is None or grad is None:
                    continue

                state = self.state[p]
                shape = grad.shape
                rank = len(shape)

                # State initialization
                if len(state) == 0:
                    state['step'] = 0
                    state['momentum_buffer'] = 0.
                    _add_initial_accumulators(state, grad)

                if grad.is_sparse:
                    # the update is non-linear so indices must be unique
                    grad.coalesce()
                    grad_indices = grad._indices()
                    grad_values = grad._values()

                    # Transform update_values into sparse tensor
                    def make_sparse(values):
                        constructor = grad.new
                        if grad_indices.dim() == 0 or values.dim() == 0:
                            return constructor().resize_as_(grad)
                        return constructor(grad_indices, values, grad.size())

                    acc = state[_key(0)]
                    update_values = _compute_sparse_update(beta, acc, grad_values, grad_indices)

                    self._update_sparse_accumulator(beta, acc, make_sparse(update_values))

                    # Add small amount for numerical stability
                    update_values.add_(eps).rsqrt_().mul_(grad_values)

                    update = make_sparse(update_values)
                else:
                    # Get previous accumulators mu_{t-1}
                    if rank > 1:
                        acc_list = [state[_key(i)] for i in range(rank)]
                    else:
                        acc_list = [state[_key(0)]]

                    # Get update from accumulators and gradients
                    update = _compute_update(beta, acc_list, grad)

                    # Update accumulators.
                    self._update_accumulator(beta, acc_list, update)

                    # Add small amount for numerical stability
                    update.add_(eps).rsqrt_().mul_(grad)

                    if momentum > 0.:
                        m = state['momentum_buffer']
                        update.mul_(1. - momentum).add_(m, alpha=momentum)
                        state['momentum_buffer'] = update.detach()

                p.sub_(update, alpha=group['lr'])
                state['step'] += 1
        return loss

    def _update_accumulator(self, beta, acc_list, update):
        for i, acc in enumerate(acc_list):
            nu_max = _max_reduce_except_dim(update, i)
            if beta > 0.:
                torch.max(acc, nu_max, out=acc)
            else:
                # No need to compare - nu_max is bigger because of grad ** 2
                acc.copy_(nu_max)

    def _update_sparse_accumulator(self, beta, acc, update):
        nu_max = _max_reduce_except_dim(update.to_dense(), 0).squeeze()
        if beta > 0.:
            torch.max(acc, nu_max, out=acc)
        else:
            # No need to compare - nu_max is bigger because of grad ** 2
            acc.copy_(nu_max)

def _compute_sparse_update(beta, acc, grad_values, grad_indices):
    # In the sparse case, a single accumulator is used.
    update_values = torch.gather(acc, 0, grad_indices[0])
    if beta > 0.:
        update_values.mul_(beta)
    update_values.addcmul_(grad_values, grad_values, value=1. - beta)
    return update_values

def _compute_update(beta, acc_list, grad):
    rank = len(acc_list)
    update = acc_list[0].clone()
    for i in range(1, rank):
        # We rely on broadcasting to get the proper end shape.
        update = torch.min(update, acc_list[i])
    if beta > 0.:
        update.mul_(beta)
    update.addcmul_(grad, grad, value=1. - beta)

    return update

def _key(i):
    # Returns key used for accessing accumulators
    return 'accumulator_' + str(i)

def _add_initial_accumulators(state, grad):
    # Creates initial accumulators. For a dense tensor of shape (n1, n2, n3),
    # then our initial accumulators are of shape (n1, 1, 1), (1, n2, 1) and
    # (1, 1, n3). For a sparse tensor of shape (n, *), we use a single
    # accumulator of shape (n,).
    shape = grad.shape
    rank = len(shape)
    defaults = {'device': grad.device, 'dtype': grad.dtype}
    acc = {}

    if grad.is_sparse:
        acc[_key(0)] = torch.zeros(shape[0], **defaults)
    elif rank == 0:
        # The scalar case is handled separately
        acc[_key(0)] = torch.zeros(shape, **defaults)
    else:
        for i in range(rank):
            acc_shape = [1] * i + [shape[i]] + [1] * (rank - 1 - i)
            acc[_key(i)] = torch.zeros(acc_shape, **defaults)

    state.update(acc)

def _max_reduce_except_dim(tensor, dim):
    # Computes max along all dimensions except the given dim.
    # If tensor is a scalar, it returns tensor.
    rank = len(tensor.shape)
    result = tensor
    if rank > 0:
        assert dim < rank
        for d in range(rank):
            if d != dim:
                result = result.max(dim=d, keepdim=True).values
    return result
